{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "5c873a6c-45a5-405f-b1d0-38f4f0643b4e",
   "metadata": {},
   "source": [
    "## ChIP-Seq and RNA-Seq\n",
    "\n",
    "We now want to combine our ChIP-Seq and RNA-Seq data, using pandas."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "d757c063-198d-46f7-9bf6-835e13f57c56",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3317628c-549e-4071-b06f-5bbc763b18b2",
   "metadata": {},
   "source": [
    "### Gene annotations\n",
    "\n",
    "We can read our GFF file of gene annotations. It's just a text table.\n",
    "\n",
    "There are no headers, though. The columns of the file are:\n",
    "```\n",
    "['seqid', 'source', 'type', 'start', 'end', 'score', 'strand', 'phase', 'attributes']\n",
    "```"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "8805e4d1-0611-4959-95b8-6f949a564a38",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>telomere</td>\n",
       "      <td>1</td>\n",
       "      <td>801</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=TEL01L;Name=TEL01L;Note=Telomeric%20region%...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>X_element</td>\n",
       "      <td>337</td>\n",
       "      <td>801</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_X_element</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>X_element_combinatorial_repeat</td>\n",
       "      <td>63</td>\n",
       "      <td>336</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_X_element_combinator...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>telomeric_repeat</td>\n",
       "      <td>1</td>\n",
       "      <td>62</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>Parent=TEL01L;Name=TEL01L_telomeric_repeat_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>telomere</td>\n",
       "      <td>1</td>\n",
       "      <td>801</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=TEL01L_telomere;Name=TEL01L_telomere;Parent...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid source                            type  start  end score strand phase  \\\n",
       "0  chrI    SGD                        telomere      1  801     .      -     .   \n",
       "1  chrI    SGD                       X_element    337  801     .      -     .   \n",
       "2  chrI    SGD  X_element_combinatorial_repeat     63  336     .      -     .   \n",
       "3  chrI    SGD                telomeric_repeat      1   62     .      -     .   \n",
       "4  chrI    SGD                        telomere      1  801     .      -     .   \n",
       "\n",
       "                                          attributes  \n",
       "0  ID=TEL01L;Name=TEL01L;Note=Telomeric%20region%...  \n",
       "1                Parent=TEL01L;Name=TEL01L_X_element  \n",
       "2  Parent=TEL01L;Name=TEL01L_X_element_combinator...  \n",
       "3       Parent=TEL01L;Name=TEL01L_telomeric_repeat_1  \n",
       "4  ID=TEL01L_telomere;Name=TEL01L_telomere;Parent...  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = pd.read_csv(\"~/shared/MCB280A_data/S288C_R64-3-1/saccharomyces_cerevisiae_R64-3-1_20210421.gff\",\n",
    "                    delimiter='\\t', header=None,\n",
    "                   names=['seqid', 'source', 'type', 'start', 'end',\n",
    "                          'score', 'strand', 'phase', 'attributes'])\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a601f820-0174-4f80-bf67-13637611963e",
   "metadata": {},
   "source": [
    "There are many different \"types\" of entries in the file. We can use the `value_counts()` method on the `type` column to see all of them."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2a0ff8f7-7f19-405b-84fd-b10329788ca1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "mRNA                                  11119\n",
       "CDS                                    7066\n",
       "gene                                   6607\n",
       "noncoding_exon                          489\n",
       "long_terminal_repeat                    384\n",
       "intron                                  378\n",
       "tRNA_gene                               299\n",
       "tRNA                                    299\n",
       "transposable_element                     91\n",
       "transposable_element_gene                91\n",
       "region                                   78\n",
       "snoRNA                                   77\n",
       "snoRNA_gene                              77\n",
       "telomere                                 64\n",
       "LTR_retrotransposon                      50\n",
       "plus_1_translational_frameshift          47\n",
       "X_element                                32\n",
       "centromere                               32\n",
       "telomeric_repeat                         31\n",
       "X_element_combinatorial_repeat           28\n",
       "rRNA                                     24\n",
       "five_prime_UTR_intron                    24\n",
       "rRNA_gene                                24\n",
       "ncRNA_gene                               23\n",
       "ncRNA                                    23\n",
       "Y_prime_element                          19\n",
       "centromere_DNA_Element_III               16\n",
       "centromere_DNA_Element_II                16\n",
       "centromere_DNA_Element_I                 16\n",
       "blocked_reading_frame                    12\n",
       "pseudogene                               12\n",
       "uORF                                     11\n",
       "internal_transcribed_spacer_region        8\n",
       "external_transcribed_spacer_region        8\n",
       "matrix_attachment_site                    8\n",
       "origin_of_replication                     8\n",
       "snRNA                                     6\n",
       "snRNA_gene                                6\n",
       "silent_mating_type_cassette_array         4\n",
       "X_region                                  3\n",
       "Y_region                                  3\n",
       "Z1_region                                 3\n",
       "non_transcribed_region                    3\n",
       "W_region                                  2\n",
       "Z2_region                                 2\n",
       "mating_type_region                        2\n",
       "intein_encoding_region                    2\n",
       "telomerase_RNA                            1\n",
       "telomerase_RNA_gene                       1\n",
       "Name: type, dtype: int64"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes['type'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8ee3b4ef-3a87-431d-b6f8-53d51ab78e4e",
   "metadata": {},
   "source": [
    "We only want the `\"gene\"` type of entry and so we'll pick out just these rows out of the data frame."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "df1648dc-cf0f-49bc-af2c-880fdfdc382a",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source  type  start   end score strand phase  \\\n",
       "5   chrI    SGD  gene    335   649     .      +     .   \n",
       "8   chrI    SGD  gene    538   792     .      +     .   \n",
       "11  chrI    SGD  gene   1807  2169     .      -     .   \n",
       "14  chrI    SGD  gene   2480  2707     .      +     .   \n",
       "17  chrI    SGD  gene   7235  9016     .      -     .   \n",
       "\n",
       "                                           attributes  \n",
       "5   ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...  \n",
       "8   ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  \n",
       "11  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...  \n",
       "14  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  \n",
       "17  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...  "
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes = genes[genes['type'] == \"gene\"]\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f9e6ac91-d35c-412b-bfa4-327a705f7b09",
   "metadata": {},
   "source": [
    "The systematic name of the gene is buried inside the `attributes` column. This column has a bunch of labeled data in the form _key_=_value_.\n",
    "\n",
    "We can extract it using the `.str.split()` method and then use `.str.replace()` to get rid of the wanted `ID=` part of it to create a `name` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "22dd8115-e211-417b-8c4a-d252e177ff27",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source  type  start   end score strand phase  \\\n",
       "5   chrI    SGD  gene    335   649     .      +     .   \n",
       "8   chrI    SGD  gene    538   792     .      +     .   \n",
       "11  chrI    SGD  gene   1807  2169     .      -     .   \n",
       "14  chrI    SGD  gene   2480  2707     .      +     .   \n",
       "17  chrI    SGD  gene   7235  9016     .      -     .   \n",
       "\n",
       "                                           attributes       name  \n",
       "5   ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W  \n",
       "8   ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A  \n",
       "11  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C  \n",
       "14  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A  \n",
       "17  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C  "
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes['name'] = genes['attributes'].str.split(';').str[0].str.replace(\"ID=\", \"\")\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7148e8b3-345e-4339-ae58-61e0c8ef1ab8",
   "metadata": {},
   "source": [
    "The data table lists the coordinates of genes, but we want to look for ChIP-Seq peaks in the promoter regions. For a `+` strand gene, the promoter is to the left (smaller coordinate numbers) and for a `-` strand gene it is to the right.\n",
    "\n",
    "We'll use a 1 kb window here.\n",
    "\n",
    "For a `+` strand gene, the promoter is _start_-1000 to _start_-1.\n",
    "\n",
    "For a `-` strand gene, the promoter is _end_+1 to _end_+1000.\n",
    "\n",
    "We can use the `np.where(...)` function from `numpy` to handle this situation. We'll compute the starting position for each promoter in the `prmstart` column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "c8a5d6d6-41f9-4112-8de8-20b1b2c6fe17",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>prmstart</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>-665</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>-462</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>2170</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>1480</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>9017</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source  type  start   end score strand phase  \\\n",
       "5   chrI    SGD  gene    335   649     .      +     .   \n",
       "8   chrI    SGD  gene    538   792     .      +     .   \n",
       "11  chrI    SGD  gene   1807  2169     .      -     .   \n",
       "14  chrI    SGD  gene   2480  2707     .      +     .   \n",
       "17  chrI    SGD  gene   7235  9016     .      -     .   \n",
       "\n",
       "                                           attributes       name  prmstart  \n",
       "5   ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W      -665  \n",
       "8   ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A      -462  \n",
       "11  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C      2170  \n",
       "14  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A      1480  \n",
       "17  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C      9017  "
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import numpy as np\n",
    "genes['prmstart'] = np.where(genes['strand'] == \"+\", \n",
    "         genes['start'] - 1000,\n",
    "         genes['end'] + 1)\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8b9bd58f-c759-4e7a-b8a4-72e5af2c3719",
   "metadata": {},
   "source": [
    "We'll be sure to check our coordinate math for genes on each strand.\n",
    "\n",
    "Next, we'll make a `prmend` column that is 1000 bases after the start. We don't need to do anything different based on the strand here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "5eceab74-15cb-4604-9a67-7f9c4fe4ac87",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>prmstart</th>\n",
       "      <th>prmend</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>-665</td>\n",
       "      <td>334</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>-462</td>\n",
       "      <td>537</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>2170</td>\n",
       "      <td>3169</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>1480</td>\n",
       "      <td>2479</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>9017</td>\n",
       "      <td>10016</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source  type  start   end score strand phase  \\\n",
       "5   chrI    SGD  gene    335   649     .      +     .   \n",
       "8   chrI    SGD  gene    538   792     .      +     .   \n",
       "11  chrI    SGD  gene   1807  2169     .      -     .   \n",
       "14  chrI    SGD  gene   2480  2707     .      +     .   \n",
       "17  chrI    SGD  gene   7235  9016     .      -     .   \n",
       "\n",
       "                                           attributes       name  prmstart  \\\n",
       "5   ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W      -665   \n",
       "8   ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A      -462   \n",
       "11  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C      2170   \n",
       "14  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A      1480   \n",
       "17  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C      9017   \n",
       "\n",
       "    prmend  \n",
       "5      334  \n",
       "8      537  \n",
       "11    3169  \n",
       "14    2479  \n",
       "17   10016  "
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes['prmend'] = genes['prmstart'] + 999\n",
    "genes.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e99037a1-cfd5-4f74-b4bf-f989e163c0e4",
   "metadata": {},
   "source": [
    "### ChIP-Seq Peaks\n",
    "\n",
    "Now, we'll read in the table of ChIP-Seq peaks."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "f04b18af-cfb3-4df4-99fa-c5d59ff1274b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>chr</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>length</th>\n",
       "      <th>abs_summit</th>\n",
       "      <th>pileup</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>23546</td>\n",
       "      <td>23755</td>\n",
       "      <td>210</td>\n",
       "      <td>23733</td>\n",
       "      <td>50</td>\n",
       "      <td>9.64499</td>\n",
       "      <td>2.65601</td>\n",
       "      <td>7.55540</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>71056</td>\n",
       "      <td>71337</td>\n",
       "      <td>282</td>\n",
       "      <td>71176</td>\n",
       "      <td>65</td>\n",
       "      <td>7.96768</td>\n",
       "      <td>2.12219</td>\n",
       "      <td>5.99254</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>71873</td>\n",
       "      <td>73595</td>\n",
       "      <td>1723</td>\n",
       "      <td>72380</td>\n",
       "      <td>89</td>\n",
       "      <td>18.71100</td>\n",
       "      <td>2.97030</td>\n",
       "      <td>16.18150</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>101112</td>\n",
       "      <td>101311</td>\n",
       "      <td>200</td>\n",
       "      <td>101291</td>\n",
       "      <td>45</td>\n",
       "      <td>7.47097</td>\n",
       "      <td>2.39562</td>\n",
       "      <td>5.53262</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>139663</td>\n",
       "      <td>141017</td>\n",
       "      <td>1355</td>\n",
       "      <td>140367</td>\n",
       "      <td>103</td>\n",
       "      <td>9.85628</td>\n",
       "      <td>1.96375</td>\n",
       "      <td>7.75715</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_5</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "    chr   start     end  length  abs_summit  pileup  -log10(pvalue)  \\\n",
       "0  chrI   23546   23755     210       23733      50         9.64499   \n",
       "1  chrI   71056   71337     282       71176      65         7.96768   \n",
       "2  chrI   71873   73595    1723       72380      89        18.71100   \n",
       "3  chrI  101112  101311     200      101291      45         7.47097   \n",
       "4  chrI  139663  141017    1355      140367     103         9.85628   \n",
       "\n",
       "   fold_enrichment  -log10(qvalue)                      name  \n",
       "0          2.65601         7.55540  Hsf1_ChIP_heatshk_peak_1  \n",
       "1          2.12219         5.99254  Hsf1_ChIP_heatshk_peak_2  \n",
       "2          2.97030        16.18150  Hsf1_ChIP_heatshk_peak_3  \n",
       "3          2.39562         5.53262  Hsf1_ChIP_heatshk_peak_4  \n",
       "4          1.96375         7.75715  Hsf1_ChIP_heatshk_peak_5  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "peaks = pd.read_csv(\"~/Hsf1/ChIP-Seq/macs2/Hsf1_ChIP_heatshk_peaks.xls\",\n",
    "                    delimiter = '\\t', comment = '#')\n",
    "peaks.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88b489f6-1608-4f8f-9a63-4526bf0bbf10",
   "metadata": {},
   "source": [
    "At this point we need to loop over each peak and find out which promoter(s) it affects.\n",
    "\n",
    "We'll do this in a multi-step process.\n",
    "1. Generate a dictionary where keys are gene names and values are associated ChIP-Seq peak names. There won't be an entry for every gene.\n",
    "1. Convert this dictionary into a pandas `Series` and merge it in to a new column of the genes data frame, holding peak names. It will have many `NaN` entries for \"missing\" data.\n",
    "1. Merge the whole table of peaks (with enrichment and p-values) into the gene table based on the names.\n",
    "\n",
    "First, we will loop over each peak, using the `.itertuples()` method. In class, we'll just use the top 10 peaks by significance."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "f05af3b5-61fd-4f67-9705-e735340df743",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'YGR210C': 'Hsf1_ChIP_heatshk_peak_389',\n",
       " 'YGR211W': 'Hsf1_ChIP_heatshk_peak_389',\n",
       " 'YBR101C': 'Hsf1_ChIP_heatshk_peak_69',\n",
       " 'YGR142W': 'Hsf1_ChIP_heatshk_peak_372',\n",
       " 'YPL250C': 'Hsf1_ChIP_heatshk_peak_802',\n",
       " 'YPL250W-A': 'Hsf1_ChIP_heatshk_peak_802',\n",
       " 'YLR162W-A': 'Hsf1_ChIP_heatshk_peak_559',\n",
       " 'YDR258C': 'Hsf1_ChIP_heatshk_peak_167',\n",
       " 'YNL007C': 'Hsf1_ChIP_heatshk_peak_708',\n",
       " 'YNL006W': 'Hsf1_ChIP_heatshk_peak_708',\n",
       " 'YDR214W': 'Hsf1_ChIP_heatshk_peak_162',\n",
       " 'YER103W': 'Hsf1_ChIP_heatshk_peak_276'}"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_peaks = {}\n",
    "top_peaks = peaks.sort_values('fold_enrichment', ascending=False).head(10)\n",
    "for peak in top_peaks.itertuples():\n",
    "    for gene in genes.itertuples():\n",
    "        if (peak.chr == gene.seqid) and (peak.abs_summit >= gene.prmstart) and (peak.abs_summit <= gene.prmend):\n",
    "            gene_peaks[gene.name] = peak.name\n",
    "gene_peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b811e890-cf4f-43ae-9ed8-d736422caea9",
   "metadata": {
    "tags": []
   },
   "source": [
    "Now we need to convert this into a `Series` and give it a name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "bcba89f4-cf96-4989-a194-908862f80e26",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "YGR210C      Hsf1_ChIP_heatshk_peak_389\n",
       "YGR211W      Hsf1_ChIP_heatshk_peak_389\n",
       "YBR101C       Hsf1_ChIP_heatshk_peak_69\n",
       "YGR142W      Hsf1_ChIP_heatshk_peak_372\n",
       "YPL250C      Hsf1_ChIP_heatshk_peak_802\n",
       "YPL250W-A    Hsf1_ChIP_heatshk_peak_802\n",
       "YLR162W-A    Hsf1_ChIP_heatshk_peak_559\n",
       "YDR258C      Hsf1_ChIP_heatshk_peak_167\n",
       "YNL007C      Hsf1_ChIP_heatshk_peak_708\n",
       "YNL006W      Hsf1_ChIP_heatshk_peak_708\n",
       "YDR214W      Hsf1_ChIP_heatshk_peak_162\n",
       "YER103W      Hsf1_ChIP_heatshk_peak_276\n",
       "Name: peak, dtype: object"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gene_peaks = pd.Series(gene_peaks, name='peak')\n",
    "gene_peaks"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "066c5ce0-d406-4277-993f-48a16be3eb37",
   "metadata": {},
   "source": [
    "We can merge our named `Series` into the data frame of genes. We want to match up the `name` column in the genes table with the \"index\" of the `Series."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "77447701-98f6-4f5f-b3df-46fa162213fb",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>prmstart</th>\n",
       "      <th>prmend</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>-665</td>\n",
       "      <td>334</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>-462</td>\n",
       "      <td>537</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>2170</td>\n",
       "      <td>3169</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>1480</td>\n",
       "      <td>2479</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>9017</td>\n",
       "      <td>10016</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   seqid source  type  start   end score strand phase  \\\n",
       "5   chrI    SGD  gene    335   649     .      +     .   \n",
       "8   chrI    SGD  gene    538   792     .      +     .   \n",
       "11  chrI    SGD  gene   1807  2169     .      -     .   \n",
       "14  chrI    SGD  gene   2480  2707     .      +     .   \n",
       "17  chrI    SGD  gene   7235  9016     .      -     .   \n",
       "\n",
       "                                           attributes       name  prmstart  \\\n",
       "5   ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W      -665   \n",
       "8   ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A      -462   \n",
       "11  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C      2170   \n",
       "14  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A      1480   \n",
       "17  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C      9017   \n",
       "\n",
       "    prmend peak  \n",
       "5      334  NaN  \n",
       "8      537  NaN  \n",
       "11    3169  NaN  \n",
       "14    2479  NaN  \n",
       "17   10016  NaN  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes2 = pd.merge(genes, gene_peaks,\n",
    "                  left_on='name', right_index=True, how='left')\n",
    "genes2.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03016b1e-825d-42f3-8f6d-77fc2410f15e",
   "metadata": {},
   "source": [
    "To make sure this worked, check on the row for the gene `YAL005C`, which does have a peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "0c04c971-b863-4bf8-a613-cd60aa904047",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name</th>\n",
       "      <th>prmstart</th>\n",
       "      <th>prmend</th>\n",
       "      <th>peak</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>10702</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>913503</td>\n",
       "      <td>914738</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...</td>\n",
       "      <td>YGR210C</td>\n",
       "      <td>914739</td>\n",
       "      <td>915738</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        seqid source  type   start     end score strand phase  \\\n",
       "10702  chrVII    SGD  gene  913503  914738     .      -     .   \n",
       "\n",
       "                                              attributes     name  prmstart  \\\n",
       "10702  ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...  YGR210C    914739   \n",
       "\n",
       "       prmend                        peak  \n",
       "10702  915738  Hsf1_ChIP_heatshk_peak_389  "
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes2[genes2['name'] == \"YGR210C\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "89dc3859-eaf3-4444-943d-390609e74a19",
   "metadata": {},
   "source": [
    "Now, we will merge in the peaks table by matching up the `peak` column with the `name` column in the peaks table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "4504c9fe-c98a-4fa2-9522-afd6068a6973",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name_x</th>\n",
       "      <th>...</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>length</th>\n",
       "      <th>abs_summit</th>\n",
       "      <th>pileup</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid source  type  start_x  end_x score strand phase  \\\n",
       "0  chrI    SGD  gene      335    649     .      +     .   \n",
       "1  chrI    SGD  gene      538    792     .      +     .   \n",
       "2  chrI    SGD  gene     1807   2169     .      -     .   \n",
       "3  chrI    SGD  gene     2480   2707     .      +     .   \n",
       "4  chrI    SGD  gene     7235   9016     .      -     .   \n",
       "\n",
       "                                          attributes     name_x  ...  chr  \\\n",
       "0  ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W  ...  NaN   \n",
       "1  ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A  ...  NaN   \n",
       "2  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C  ...  NaN   \n",
       "3  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A  ...  NaN   \n",
       "4  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C  ...  NaN   \n",
       "\n",
       "   start_y end_y length  abs_summit  pileup  -log10(pvalue)  fold_enrichment  \\\n",
       "0      NaN   NaN    NaN         NaN     NaN             NaN              NaN   \n",
       "1      NaN   NaN    NaN         NaN     NaN             NaN              NaN   \n",
       "2      NaN   NaN    NaN         NaN     NaN             NaN              NaN   \n",
       "3      NaN   NaN    NaN         NaN     NaN             NaN              NaN   \n",
       "4      NaN   NaN    NaN         NaN     NaN             NaN              NaN   \n",
       "\n",
       "   -log10(qvalue)  name_y  \n",
       "0             NaN     NaN  \n",
       "1             NaN     NaN  \n",
       "2             NaN     NaN  \n",
       "3             NaN     NaN  \n",
       "4             NaN     NaN  \n",
       "\n",
       "[5 rows x 23 columns]"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes3 = pd.merge(genes2, peaks,\n",
    "                  left_on='peak', right_on='name', how='left')\n",
    "genes3.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4034399e-cde8-4a4a-bcbe-54e8673a2700",
   "metadata": {},
   "source": [
    "Again, we'll need to check on a row that has a peak to be sure it worked."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "8f99613a-de6c-488d-a82b-03b6fa000eae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name_x</th>\n",
       "      <th>...</th>\n",
       "      <th>chr</th>\n",
       "      <th>start_y</th>\n",
       "      <th>end_y</th>\n",
       "      <th>length</th>\n",
       "      <th>abs_summit</th>\n",
       "      <th>pileup</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name_y</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2547</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>913503</td>\n",
       "      <td>914738</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...</td>\n",
       "      <td>YGR210C</td>\n",
       "      <td>...</td>\n",
       "      <td>chrVII</td>\n",
       "      <td>914726.0</td>\n",
       "      <td>915341.0</td>\n",
       "      <td>616.0</td>\n",
       "      <td>915027.0</td>\n",
       "      <td>321.0</td>\n",
       "      <td>235.086</td>\n",
       "      <td>12.8594</td>\n",
       "      <td>229.177</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_389</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows  23 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid source  type  start_x   end_x score strand phase  \\\n",
       "2547  chrVII    SGD  gene   913503  914738     .      -     .   \n",
       "\n",
       "                                             attributes   name_x  ...     chr  \\\n",
       "2547  ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...  YGR210C  ...  chrVII   \n",
       "\n",
       "       start_y     end_y length  abs_summit  pileup  -log10(pvalue)  \\\n",
       "2547  914726.0  915341.0  616.0    915027.0   321.0         235.086   \n",
       "\n",
       "      fold_enrichment  -log10(qvalue)                      name_y  \n",
       "2547          12.8594         229.177  Hsf1_ChIP_heatshk_peak_389  \n",
       "\n",
       "[1 rows x 23 columns]"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes3[genes3['name_x'] == \"YGR210C\"]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "deb8c5e9-3a96-4cd9-92d8-69c4067ee62b",
   "metadata": {},
   "source": [
    "### RNA-Seq data\n",
    "\n",
    "Finally, we're ready to read in the table of RNA-Seq results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "6125214d-2b4d-4ce6-8aab-efe1f0fb3387",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>Q0010</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q0017</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q0032</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q0045</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Q0050</th>\n",
       "      <td>0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "       baseMean  log2FoldChange  lfcSE  stat  pvalue  padj\n",
       "Q0010       0.0             NaN    NaN   NaN     NaN   NaN\n",
       "Q0017       0.0             NaN    NaN   NaN     NaN   NaN\n",
       "Q0032       0.0             NaN    NaN   NaN     NaN   NaN\n",
       "Q0045       0.0             NaN    NaN   NaN     NaN   NaN\n",
       "Q0050       0.0             NaN    NaN   NaN     NaN   NaN"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results = pd.read_csv(\"~/Hsf1/RNA-Seq/full.results.csv\",\n",
    "                      index_col = 0)\n",
    "results.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d277ca7f-6725-4c8e-ac32-e528bcd0c595",
   "metadata": {
    "tags": []
   },
   "source": [
    "Again we can pull out some statistically significant genes to be sure it worked. Notice that almost every significant gene is down-regulated."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "e1ab8ba9-8b21-4599-bf74-1d1a9750ee4c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>YAL005C</th>\n",
       "      <td>4973.736731</td>\n",
       "      <td>-2.831197</td>\n",
       "      <td>0.269331</td>\n",
       "      <td>-10.511978</td>\n",
       "      <td>7.608047e-26</td>\n",
       "      <td>1.211011e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR067C</th>\n",
       "      <td>6601.912986</td>\n",
       "      <td>-1.598963</td>\n",
       "      <td>0.373009</td>\n",
       "      <td>-4.286660</td>\n",
       "      <td>1.813795e-05</td>\n",
       "      <td>7.698954e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YBR101C</th>\n",
       "      <td>3528.512278</td>\n",
       "      <td>-1.215401</td>\n",
       "      <td>0.289295</td>\n",
       "      <td>-4.201254</td>\n",
       "      <td>2.654410e-05</td>\n",
       "      <td>1.056289e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YDR214W</th>\n",
       "      <td>3436.341557</td>\n",
       "      <td>-1.875259</td>\n",
       "      <td>0.301106</td>\n",
       "      <td>-6.227899</td>\n",
       "      <td>4.727310e-10</td>\n",
       "      <td>3.762348e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YDR216W</th>\n",
       "      <td>1048.683844</td>\n",
       "      <td>-1.567399</td>\n",
       "      <td>0.364029</td>\n",
       "      <td>-4.305702</td>\n",
       "      <td>1.664567e-05</td>\n",
       "      <td>7.570215e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YDR258C</th>\n",
       "      <td>1290.992019</td>\n",
       "      <td>-1.464184</td>\n",
       "      <td>0.325875</td>\n",
       "      <td>-4.493090</td>\n",
       "      <td>7.019698e-06</td>\n",
       "      <td>3.438032e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YFR036W-A</th>\n",
       "      <td>64.658953</td>\n",
       "      <td>2.234114</td>\n",
       "      <td>0.564499</td>\n",
       "      <td>3.957693</td>\n",
       "      <td>7.567706e-05</td>\n",
       "      <td>2.676866e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YGR142W</th>\n",
       "      <td>420.520506</td>\n",
       "      <td>-1.922651</td>\n",
       "      <td>0.368085</td>\n",
       "      <td>-5.223394</td>\n",
       "      <td>1.756733e-07</td>\n",
       "      <td>1.016829e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YJR046W</th>\n",
       "      <td>1326.102238</td>\n",
       "      <td>-1.134671</td>\n",
       "      <td>0.313169</td>\n",
       "      <td>-3.623187</td>\n",
       "      <td>2.909957e-04</td>\n",
       "      <td>9.263849e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLL024C</th>\n",
       "      <td>47370.552792</td>\n",
       "      <td>-4.455708</td>\n",
       "      <td>0.259150</td>\n",
       "      <td>-17.193563</td>\n",
       "      <td>2.967229e-66</td>\n",
       "      <td>1.889235e-62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLL026W</th>\n",
       "      <td>8418.821531</td>\n",
       "      <td>-1.736765</td>\n",
       "      <td>0.311697</td>\n",
       "      <td>-5.571970</td>\n",
       "      <td>2.518744e-08</td>\n",
       "      <td>1.603684e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YLR216C</th>\n",
       "      <td>7870.274759</td>\n",
       "      <td>-1.125082</td>\n",
       "      <td>0.275573</td>\n",
       "      <td>-4.082694</td>\n",
       "      <td>4.451663e-05</td>\n",
       "      <td>1.667279e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YMR186W</th>\n",
       "      <td>26117.391850</td>\n",
       "      <td>-3.335366</td>\n",
       "      <td>0.278353</td>\n",
       "      <td>-11.982493</td>\n",
       "      <td>4.389232e-33</td>\n",
       "      <td>1.397312e-29</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YNL007C</th>\n",
       "      <td>1067.866637</td>\n",
       "      <td>-2.608539</td>\n",
       "      <td>0.292907</td>\n",
       "      <td>-8.905696</td>\n",
       "      <td>5.305227e-19</td>\n",
       "      <td>6.755676e-16</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YNL064C</th>\n",
       "      <td>3276.125168</td>\n",
       "      <td>-1.561059</td>\n",
       "      <td>0.278786</td>\n",
       "      <td>-5.599498</td>\n",
       "      <td>2.149731e-08</td>\n",
       "      <td>1.520815e-05</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YOR027W</th>\n",
       "      <td>2243.863806</td>\n",
       "      <td>-2.413686</td>\n",
       "      <td>0.281530</td>\n",
       "      <td>-8.573448</td>\n",
       "      <td>1.004316e-17</td>\n",
       "      <td>1.065747e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YOR298C-A</th>\n",
       "      <td>8347.855176</td>\n",
       "      <td>-2.097829</td>\n",
       "      <td>0.277320</td>\n",
       "      <td>-7.564657</td>\n",
       "      <td>3.888893e-14</td>\n",
       "      <td>3.537226e-11</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPL106C</th>\n",
       "      <td>19737.734943</td>\n",
       "      <td>-0.966315</td>\n",
       "      <td>0.263985</td>\n",
       "      <td>-3.660499</td>\n",
       "      <td>2.517246e-04</td>\n",
       "      <td>8.435425e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPL240C</th>\n",
       "      <td>4047.260117</td>\n",
       "      <td>-3.465542</td>\n",
       "      <td>0.329240</td>\n",
       "      <td>-10.525870</td>\n",
       "      <td>6.565191e-26</td>\n",
       "      <td>1.211011e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>YPR158W</th>\n",
       "      <td>517.786713</td>\n",
       "      <td>-1.721643</td>\n",
       "      <td>0.339918</td>\n",
       "      <td>-5.064884</td>\n",
       "      <td>4.086495e-07</td>\n",
       "      <td>2.168226e-04</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "               baseMean  log2FoldChange     lfcSE       stat        pvalue  \\\n",
       "YAL005C     4973.736731       -2.831197  0.269331 -10.511978  7.608047e-26   \n",
       "YBR067C     6601.912986       -1.598963  0.373009  -4.286660  1.813795e-05   \n",
       "YBR101C     3528.512278       -1.215401  0.289295  -4.201254  2.654410e-05   \n",
       "YDR214W     3436.341557       -1.875259  0.301106  -6.227899  4.727310e-10   \n",
       "YDR216W     1048.683844       -1.567399  0.364029  -4.305702  1.664567e-05   \n",
       "YDR258C     1290.992019       -1.464184  0.325875  -4.493090  7.019698e-06   \n",
       "YFR036W-A     64.658953        2.234114  0.564499   3.957693  7.567706e-05   \n",
       "YGR142W      420.520506       -1.922651  0.368085  -5.223394  1.756733e-07   \n",
       "YJR046W     1326.102238       -1.134671  0.313169  -3.623187  2.909957e-04   \n",
       "YLL024C    47370.552792       -4.455708  0.259150 -17.193563  2.967229e-66   \n",
       "YLL026W     8418.821531       -1.736765  0.311697  -5.571970  2.518744e-08   \n",
       "YLR216C     7870.274759       -1.125082  0.275573  -4.082694  4.451663e-05   \n",
       "YMR186W    26117.391850       -3.335366  0.278353 -11.982493  4.389232e-33   \n",
       "YNL007C     1067.866637       -2.608539  0.292907  -8.905696  5.305227e-19   \n",
       "YNL064C     3276.125168       -1.561059  0.278786  -5.599498  2.149731e-08   \n",
       "YOR027W     2243.863806       -2.413686  0.281530  -8.573448  1.004316e-17   \n",
       "YOR298C-A   8347.855176       -2.097829  0.277320  -7.564657  3.888893e-14   \n",
       "YPL106C    19737.734943       -0.966315  0.263985  -3.660499  2.517246e-04   \n",
       "YPL240C     4047.260117       -3.465542  0.329240 -10.525870  6.565191e-26   \n",
       "YPR158W      517.786713       -1.721643  0.339918  -5.064884  4.086495e-07   \n",
       "\n",
       "                   padj  \n",
       "YAL005C    1.211011e-22  \n",
       "YBR067C    7.698954e-03  \n",
       "YBR101C    1.056289e-02  \n",
       "YDR214W    3.762348e-07  \n",
       "YDR216W    7.570215e-03  \n",
       "YDR258C    3.438032e-03  \n",
       "YFR036W-A  2.676866e-02  \n",
       "YGR142W    1.016829e-04  \n",
       "YJR046W    9.263849e-02  \n",
       "YLL024C    1.889235e-62  \n",
       "YLL026W    1.603684e-05  \n",
       "YLR216C    1.667279e-02  \n",
       "YMR186W    1.397312e-29  \n",
       "YNL007C    6.755676e-16  \n",
       "YNL064C    1.520815e-05  \n",
       "YOR027W    1.065747e-14  \n",
       "YOR298C-A  3.537226e-11  \n",
       "YPL106C    8.435425e-02  \n",
       "YPL240C    1.211011e-22  \n",
       "YPR158W    2.168226e-04  "
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "results[results['padj'] < 0.1]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8cef69e4-fb37-4579-b085-9e97f7fbb09c",
   "metadata": {},
   "source": [
    "Finally, we're ready to merge the results with the genes.\n",
    "\n",
    "We'll match up the `name_x` column of the genes (which was renamed because the peak table also had a name column) with the \"index\" of the results table."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "1c2a0d22-6c06-41e4-9f31-983c74856604",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name_x</th>\n",
       "      <th>...</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name_y</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>335</td>\n",
       "      <td>649</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...</td>\n",
       "      <td>YAL069W</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.478392</td>\n",
       "      <td>2.265198</td>\n",
       "      <td>4.888960</td>\n",
       "      <td>0.463329</td>\n",
       "      <td>0.643128</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>538</td>\n",
       "      <td>792</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL068W-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>1807</td>\n",
       "      <td>2169</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...</td>\n",
       "      <td>YAL068C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1.061544</td>\n",
       "      <td>-3.645244</td>\n",
       "      <td>3.762115</td>\n",
       "      <td>-0.968935</td>\n",
       "      <td>0.332578</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>2480</td>\n",
       "      <td>2707</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...</td>\n",
       "      <td>YAL067W-A</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>chrI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>7235</td>\n",
       "      <td>9016</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...</td>\n",
       "      <td>YAL067C</td>\n",
       "      <td>...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>123.571978</td>\n",
       "      <td>0.221327</td>\n",
       "      <td>0.419384</td>\n",
       "      <td>0.527743</td>\n",
       "      <td>0.597678</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "  seqid source  type  start_x  end_x score strand phase  \\\n",
       "0  chrI    SGD  gene      335    649     .      +     .   \n",
       "1  chrI    SGD  gene      538    792     .      +     .   \n",
       "2  chrI    SGD  gene     1807   2169     .      -     .   \n",
       "3  chrI    SGD  gene     2480   2707     .      +     .   \n",
       "4  chrI    SGD  gene     7235   9016     .      -     .   \n",
       "\n",
       "                                          attributes     name_x  ...  \\\n",
       "0  ID=YAL069W;Name=YAL069W;Ontology_term=GO:00036...    YAL069W  ...   \n",
       "1  ID=YAL068W-A;Name=YAL068W-A;Ontology_term=GO:0...  YAL068W-A  ...   \n",
       "2  ID=YAL068C;Name=YAL068C;gene=PAU8;Alias=PAU8,s...    YAL068C  ...   \n",
       "3  ID=YAL067W-A;Name=YAL067W-A;Ontology_term=GO:0...  YAL067W-A  ...   \n",
       "4  ID=YAL067C;Name=YAL067C;gene=SEO1;Alias=SEO1,p...    YAL067C  ...   \n",
       "\n",
       "   -log10(pvalue)  fold_enrichment -log10(qvalue) name_y    baseMean  \\\n",
       "0             NaN              NaN            NaN    NaN    0.478392   \n",
       "1             NaN              NaN            NaN    NaN    0.000000   \n",
       "2             NaN              NaN            NaN    NaN    1.061544   \n",
       "3             NaN              NaN            NaN    NaN    0.000000   \n",
       "4             NaN              NaN            NaN    NaN  123.571978   \n",
       "\n",
       "   log2FoldChange     lfcSE      stat    pvalue      padj  \n",
       "0        2.265198  4.888960  0.463329  0.643128  0.999885  \n",
       "1             NaN       NaN       NaN       NaN       NaN  \n",
       "2       -3.645244  3.762115 -0.968935  0.332578  0.999885  \n",
       "3             NaN       NaN       NaN       NaN       NaN  \n",
       "4        0.221327  0.419384  0.527743  0.597678  0.999885  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes4 = pd.merge(genes3, results,\n",
    "                  left_on='name_x', right_index=True, how='left')\n",
    "genes4.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "08e5e4a8-a8b2-4991-bc24-c1a1a7774569",
   "metadata": {},
   "source": [
    "Let's check out the ChIP-Seq genes in the RNA-Seq data."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "c1702593-1394-4389-9373-7b04f2a1116c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name_x</th>\n",
       "      <th>...</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name_y</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2547</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>913503</td>\n",
       "      <td>914738</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...</td>\n",
       "      <td>YGR210C</td>\n",
       "      <td>...</td>\n",
       "      <td>235.086</td>\n",
       "      <td>12.8594</td>\n",
       "      <td>229.177</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_389</td>\n",
       "      <td>1465.764956</td>\n",
       "      <td>-0.054162</td>\n",
       "      <td>0.293441</td>\n",
       "      <td>-0.184576</td>\n",
       "      <td>8.535615e-01</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2548</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>915241</td>\n",
       "      <td>916701</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR211W;Name=YGR211W;gene=ZPR1;Alias=ZPR1,z...</td>\n",
       "      <td>YGR211W</td>\n",
       "      <td>...</td>\n",
       "      <td>235.086</td>\n",
       "      <td>12.8594</td>\n",
       "      <td>229.177</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_389</td>\n",
       "      <td>5726.838885</td>\n",
       "      <td>-0.687502</td>\n",
       "      <td>0.293452</td>\n",
       "      <td>-2.342811</td>\n",
       "      <td>1.913905e-02</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>chrII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>443821</td>\n",
       "      <td>444693</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YBR101C;Name=YBR101C;gene=FES1;Alias=FES1;O...</td>\n",
       "      <td>YBR101C</td>\n",
       "      <td>...</td>\n",
       "      <td>158.552</td>\n",
       "      <td>10.6449</td>\n",
       "      <td>154.402</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_69</td>\n",
       "      <td>3528.512278</td>\n",
       "      <td>-1.215401</td>\n",
       "      <td>0.289295</td>\n",
       "      <td>-4.201254</td>\n",
       "      <td>2.654410e-05</td>\n",
       "      <td>0.010563</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2474</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>772454</td>\n",
       "      <td>773686</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR142W;Name=YGR142W;gene=BTN2;Alias=BTN2;O...</td>\n",
       "      <td>YGR142W</td>\n",
       "      <td>...</td>\n",
       "      <td>186.912</td>\n",
       "      <td>10.2962</td>\n",
       "      <td>182.373</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_372</td>\n",
       "      <td>420.520506</td>\n",
       "      <td>-1.922651</td>\n",
       "      <td>0.368085</td>\n",
       "      <td>-5.223394</td>\n",
       "      <td>1.756733e-07</td>\n",
       "      <td>0.000102</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6101</th>\n",
       "      <td>chrXVI</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>74309</td>\n",
       "      <td>74719</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YPL250C;Name=YPL250C;gene=ATG41;Alias=ATG41...</td>\n",
       "      <td>YPL250C</td>\n",
       "      <td>...</td>\n",
       "      <td>172.006</td>\n",
       "      <td>10.1810</td>\n",
       "      <td>167.709</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_802</td>\n",
       "      <td>1333.907690</td>\n",
       "      <td>0.020420</td>\n",
       "      <td>0.314142</td>\n",
       "      <td>0.065002</td>\n",
       "      <td>9.481720e-01</td>\n",
       "      <td>0.999885</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid source  type  start_x   end_x score strand phase  \\\n",
       "2547  chrVII    SGD  gene   913503  914738     .      -     .   \n",
       "2548  chrVII    SGD  gene   915241  916701     .      +     .   \n",
       "349    chrII    SGD  gene   443821  444693     .      -     .   \n",
       "2474  chrVII    SGD  gene   772454  773686     .      +     .   \n",
       "6101  chrXVI    SGD  gene    74309   74719     .      -     .   \n",
       "\n",
       "                                             attributes   name_x  ...  \\\n",
       "2547  ID=YGR210C;Name=YGR210C;Ontology_term=GO:00036...  YGR210C  ...   \n",
       "2548  ID=YGR211W;Name=YGR211W;gene=ZPR1;Alias=ZPR1,z...  YGR211W  ...   \n",
       "349   ID=YBR101C;Name=YBR101C;gene=FES1;Alias=FES1;O...  YBR101C  ...   \n",
       "2474  ID=YGR142W;Name=YGR142W;gene=BTN2;Alias=BTN2;O...  YGR142W  ...   \n",
       "6101  ID=YPL250C;Name=YPL250C;gene=ATG41;Alias=ATG41...  YPL250C  ...   \n",
       "\n",
       "      -log10(pvalue)  fold_enrichment -log10(qvalue)  \\\n",
       "2547         235.086          12.8594        229.177   \n",
       "2548         235.086          12.8594        229.177   \n",
       "349          158.552          10.6449        154.402   \n",
       "2474         186.912          10.2962        182.373   \n",
       "6101         172.006          10.1810        167.709   \n",
       "\n",
       "                          name_y     baseMean  log2FoldChange     lfcSE  \\\n",
       "2547  Hsf1_ChIP_heatshk_peak_389  1465.764956       -0.054162  0.293441   \n",
       "2548  Hsf1_ChIP_heatshk_peak_389  5726.838885       -0.687502  0.293452   \n",
       "349    Hsf1_ChIP_heatshk_peak_69  3528.512278       -1.215401  0.289295   \n",
       "2474  Hsf1_ChIP_heatshk_peak_372   420.520506       -1.922651  0.368085   \n",
       "6101  Hsf1_ChIP_heatshk_peak_802  1333.907690        0.020420  0.314142   \n",
       "\n",
       "          stat        pvalue      padj  \n",
       "2547 -0.184576  8.535615e-01  0.999885  \n",
       "2548 -2.342811  1.913905e-02  0.999885  \n",
       "349  -4.201254  2.654410e-05  0.010563  \n",
       "2474 -5.223394  1.756733e-07  0.000102  \n",
       "6101  0.065002  9.481720e-01  0.999885  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "genes4.sort_values('fold_enrichment', ascending=False).head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1ba8550c-792d-41fc-b3e5-17c81097fcc6",
   "metadata": {},
   "source": [
    "Of course, most genes have a `NaN` missing value for fold enrichment. We can use the `pd.isna()` function to test whether a value is `NaN` or not.\n",
    "\n",
    "Rows where `fold_enrichment` is _not_ `NaN` are genes with a potential ChIP-Seq peak."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "391eac59-bf4d-425b-9eb1-4babd9f48c76",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True     6595\n",
       "False      12\n",
       "Name: fold_enrichment, dtype: int64"
      ]
     },
     "execution_count": 35,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.isna(genes4['fold_enrichment']).value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e73ceadd-f6c7-4893-b77c-42154003263a",
   "metadata": {},
   "source": [
    "Many of these genes have significant adjusted p-values and negative log fold-changes. Let's look at this trend more rigorously by plotting the histogram of fold-changes for these groups.\n",
    "\n",
    "Start by importing matplotlib.pyplot."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "b58648fd-9921-4177-9ccc-36d8fa4e72eb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13f5bdd1-444f-4329-896b-74d22ee76da9",
   "metadata": {},
   "source": [
    "We'll make a histogram of `fold_enrichment` values for all genes."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "8d5ec819-306e-4bf9-a5b8-2f30e6613a77",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([   6.,   12.,   23.,   71.,  500., 5432.,  232.,   54.,   27.,\n",
       "          10.]),\n",
       " array([-5.21510413, -4.23470041, -3.25429669, -2.27389297, -1.29348925,\n",
       "        -0.31308553,  0.66731819,  1.64772191,  2.62812563,  3.60852935,\n",
       "         4.58893308]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 37,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAX0AAAD4CAYAAAAAczaOAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAPCElEQVR4nO3dYajdd33H8ffHxNUyDbb0tgu5YcmDPDDtptJLFuiDOeNsZovpk0Ic2sAKgZJBBUUSfTD2INAxcK5s7QhOmqJbCag0KN3MMssYdNZbrdY0Zg22a6/JmqsyjE86Ur97cH8dh+Tk3pP23nOS/N4vOPz/53t+v3O+f0o/+fM7//+5qSokSX1426QbkCSNj6EvSR0x9CWpI4a+JHXE0JekjqyedANLueGGG2rDhg2TbkOSrijPPPPMz6pq6vz6ZR/6GzZsYHZ2dtJtSNIVJcl/Dau7vCNJHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR257O/IlS5XG/Z+c2Kf/dIDd0zss3Vl80xfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0ZKfSTvJTkuSTPJpltteuTHEnyQtteNzB+X5KTSU4kuX2gfmt7n5NJHkyS5T8kSdLFXMqZ/h9U1fuqaqY93wscrapNwNH2nCSbgZ3AzcB24KEkq9qch4HdwKb22P7WD0GSNKq3sryzAzjY9g8Cdw3UH6uq16rqReAksCXJWmBNVT1VVQU8OjBHkjQGo4Z+Ad9K8kyS3a12U1WdBmjbG1t9HfDKwNy5VlvX9s+vXyDJ7iSzSWbn5+dHbFGStJRR/4jKbVV1KsmNwJEkP15k7LB1+lqkfmGx6gBwAGBmZmboGEnSpRvpTL+qTrXtGeDrwBbg1bZkQ9ueacPngPUD06eBU60+PaQuSRqTJUM/yW8medcb+8CHgR8Bh4Fdbdgu4PG2fxjYmeSaJBtZ+ML26bYEdDbJ1nbVzj0DcyRJYzDK8s5NwNfb1ZWrgX+oqn9K8l3gUJJ7gZeBuwGq6liSQ8DzwDlgT1W93t7rPuAR4FrgifaQJI3JkqFfVT8B3juk/nNg20Xm7Af2D6nPArdcepuSpOXgHbmS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0ZOfSTrEry/STfaM+vT3IkyQtte93A2H1JTiY5keT2gfqtSZ5rrz2YJMt7OJKkxVzKmf79wPGB53uBo1W1CTjanpNkM7ATuBnYDjyUZFWb8zCwG9jUHtvfUveSpEsyUugnmQbuAL44UN4BHGz7B4G7BuqPVdVrVfUicBLYkmQtsKaqnqqqAh4dmCNJGoNRz/S/AHwG+PVA7aaqOg3Qtje2+jrglYFxc622ru2fX79Akt1JZpPMzs/Pj9iiJGkpS4Z+kjuBM1X1zIjvOWydvhapX1isOlBVM1U1MzU1NeLHSpKWsnqEMbcBH03yEeAdwJokXwZeTbK2qk63pZszbfwcsH5g/jRwqtWnh9QlSWOy5Jl+Ve2rqumq2sDCF7T/WlUfBw4Du9qwXcDjbf8wsDPJNUk2svCF7dNtCehskq3tqp17BuZIksZglDP9i3kAOJTkXuBl4G6AqjqW5BDwPHAO2FNVr7c59wGPANcCT7SHJGlMLin0q+pJ4Mm2/3Ng20XG7Qf2D6nPArdcapOSpOXhHbmS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR1ZMvSTvCPJ00l+kORYkj9v9euTHEnyQtteNzBnX5KTSU4kuX2gfmuS59prDybJyhyWJGmYUc70XwM+WFXvBd4HbE+yFdgLHK2qTcDR9pwkm4GdwM3AduChJKvaez0M7AY2tcf25TsUSdJSlgz9WvCr9vTt7VHADuBgqx8E7mr7O4DHquq1qnoROAlsSbIWWFNVT1VVAY8OzJEkjcFIa/pJViV5FjgDHKmq7wA3VdVpgLa9sQ1fB7wyMH2u1da1/fPrwz5vd5LZJLPz8/OXcDiSpMWMFPpV9XpVvQ+YZuGs/ZZFhg9bp69F6sM+70BVzVTVzNTU1CgtSpJGcElX71TV/wBPsrAW/2pbsqFtz7Rhc8D6gWnTwKlWnx5SlySNyShX70wleXfbvxb4EPBj4DCwqw3bBTze9g8DO5Nck2QjC1/YPt2WgM4m2dqu2rlnYI4kaQxWjzBmLXCwXYHzNuBQVX0jyVPAoST3Ai8DdwNU1bEkh4DngXPAnqp6vb3XfcAjwLXAE+0hSRqTJUO/qn4IvH9I/efAtovM2Q/sH1KfBRb7PkCStIK8I1eSOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSNLhn6S9Um+neR4kmNJ7m/165McSfJC2143MGdfkpNJTiS5faB+a5Ln2msPJsnKHJYkaZhRzvTPAZ+qqvcAW4E9STYDe4GjVbUJONqe017bCdwMbAceSrKqvdfDwG5gU3tsX8ZjkSQtYcnQr6rTVfW9tn8WOA6sA3YAB9uwg8BdbX8H8FhVvVZVLwIngS1J1gJrquqpqirg0YE5kqQxuKQ1/SQbgPcD3wFuqqrTsPAPA3BjG7YOeGVg2lyrrWv759clSWMycugneSfwVeCTVfXLxYYOqdUi9WGftTvJbJLZ+fn5UVuUJC1hpNBP8nYWAv8rVfW1Vn61LdnQtmdafQ5YPzB9GjjV6tND6heoqgNVNVNVM1NTU6MeiyRpCaNcvRPg74HjVfX5gZcOA7va/i7g8YH6ziTXJNnIwhe2T7cloLNJtrb3vGdgjiRpDFaPMOY24BPAc0mebbXPAg8Ah5LcC7wM3A1QVceSHAKeZ+HKnz1V9Xqbdx/wCHAt8ER7SJLGZMnQr6p/Z/h6PMC2i8zZD+wfUp8FbrmUBiVJy8c7ciWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHDH1J6oihL0kdMfQlqSOGviR1xNCXpI4Y+pLUEUNfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHlgz9JF9KcibJjwZq1yc5kuSFtr1u4LV9SU4mOZHk9oH6rUmea689mCTLfziSpMWMcqb/CLD9vNpe4GhVbQKOtuck2QzsBG5ucx5KsqrNeRjYDWxqj/PfU5K0wpYM/ar6N+AX55V3AAfb/kHgroH6Y1X1WlW9CJwEtiRZC6ypqqeqqoBHB+ZIksbkza7p31RVpwHa9sZWXwe8MjBurtXWtf3z60Ml2Z1kNsns/Pz8m2xRknS+5f4id9g6fS1SH6qqDlTVTFXNTE1NLVtzktS7Nxv6r7YlG9r2TKvPAesHxk0Dp1p9ekhdkjRGbzb0DwO72v4u4PGB+s4k1yTZyMIXtk+3JaCzSba2q3buGZgjSRqT1UsNSPKPwAeAG5LMAX8GPAAcSnIv8DJwN0BVHUtyCHgeOAfsqarX21vdx8KVQNcCT7SHJGmMlgz9qvrYRV7adpHx+4H9Q+qzwC2X1J0kaVl5R64kdcTQl6SOGPqS1BFDX5I6YuhLUkcMfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOmLoS1JHlvyVTelyt2HvNyfdgnTF8Exfkjpi6EtSRwx9SeqIoS9JHTH0Jakjhr4kdcTQl6SOGPqS1BFDX5I64h250hVoUnchv/TAHRP5XC0fz/QlqSOGviR1xNCXpI4Y+pLUEUNfkjoy9qt3kmwH/hpYBXyxqh4Ydw9afv6mvXRlGGvoJ1kF/C3wh8Ac8N0kh6vq+XH2IenN8VLRK9+4z/S3ACer6icASR4DdgBXVeh71istrx7/n1qpf+jGHfrrgFcGns8Bv3f+oCS7gd3t6a+SnBhDb+N0A/CzSTcxAR53XzzutyB/8Zb7+O1hxXGHfobU6oJC1QHgwMq3MxlJZqtqZtJ9jJvH3ReP+/I07qt35oD1A8+ngVNj7kGSujXu0P8usCnJxiS/AewEDo+5B0nq1liXd6rqXJI/Bf6ZhUs2v1RVx8bZw2Xiql26WoLH3ReP+zKUqguW1CVJVynvyJWkjhj6ktQRQ3+Cknw6SSW5YdK9jEOSv0zy4yQ/TPL1JO+edE8rKcn2JCeSnEyyd9L9jEOS9Um+neR4kmNJ7p90T+OUZFWS7yf5xqR7uRhDf0KSrGfh5yhennQvY3QEuKWqfhf4T2DfhPtZMQM/OfJHwGbgY0k2T7arsTgHfKqq3gNsBfZ0ctxvuB84PukmFmPoT85fAZ9hyM1pV6uq+lZVnWtP/4OF+zSuVv//kyNV9b/AGz85clWrqtNV9b22f5aFAFw32a7GI8k0cAfwxUn3shhDfwKSfBT4aVX9YNK9TNCfAE9MuokVNOwnR7oIvzck2QC8H/jOhFsZly+wcCL36wn3sSj/MPoKSfIvwG8NeelzwGeBD4+3o/FY7Lir6vE25nMsLAN8ZZy9jdlIPzlytUryTuCrwCer6peT7melJbkTOFNVzyT5wITbWZShv0Kq6kPD6kl+B9gI/CAJLCxxfC/Jlqr67zG2uCIudtxvSLILuBPYVlf3TSLd/uRIkrezEPhfqaqvTbqfMbkN+GiSjwDvANYk+XJVfXzCfV3Am7MmLMlLwExVXfW/Rtj+gM7ngd+vqvlJ97OSkqxm4cvqbcBPWfgJkj++2u9Az8KZzEHgF1X1yQm3MxHtTP/TVXXnhFsZyjV9jdPfAO8CjiR5NsnfTbqhldK+sH7jJ0eOA4eu9sBvbgM+AXyw/Td+tp396jLhmb4kdcQzfUnqiKEvSR0x9CWpI4a+JHXE0Jekjhj6ktQRQ1+SOvJ/sBzh9qL4it8AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "plt.hist(genes4['log2FoldChange'])"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d40f8d99-f904-4a37-9484-cff4652ad535",
   "metadata": {},
   "source": [
    "Then, we'll make a similar histogram, for genes with a ChIP-Seq peak.\n",
    "\n",
    "We'll use the `range` parameter so the histograms are easier to compare."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "16d72ec4-5cfb-40b6-8373-618cb1020eb1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(array([0., 1., 2., 2., 4., 2., 0., 1., 0., 0.]),\n",
       " array([-4. , -3.2, -2.4, -1.6, -0.8,  0. ,  0.8,  1.6,  2.4,  3.2,  4. ]),\n",
       " <BarContainer object of 10 artists>)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAXQAAAD4CAYAAAD8Zh1EAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjUuMSwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy/YYfK9AAAACXBIWXMAAAsTAAALEwEAmpwYAAAQbUlEQVR4nO3df6jd9X3H8edrWYpldeSP3GGWH6Z/hLHq6g8uqeI/4uyIP1A2OkhZ63CDoFiw0NFpBUsZA0ehK5piCK20UtdSaOuCRlpHFRUWa0yT1DR2hOLmnRlJLY0NSkva9/6433Rnx3NzzknOzbn5+HzAwe+Pz/l+X15zX37zPd/v+aaqkCSd+35n2gEkSZNhoUtSIyx0SWqEhS5JjbDQJakRvzutHa9cubLWr18/rd1L0jnpxRdf/GlVzQxaN7VCX79+Pbt3757W7iXpnJTkPxda5ykXSWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1IiRCz3JsiQ/SPLYgHVJcn+SQ0n2J7l8sjElScOMc4R+J3BwgXXXARu61xbgwTPMJUka00iFnmQNcAPwxQWG3Aw8XPN2ASuSrJpQRknSCEa9U/TzwCeB8xdYvxp4tWd+rlt2uHdQki3MH8Gzbt26cXLqHWj9XY9Pbd+v3HfD1PYtna6hR+hJbgSOVNWLpxo2YNnbHoVUVduraraqZmdmBn4VgSTpNI1yyuUq4KYkrwBfB65J8tW+MXPA2p75NcBrE0koSRrJ0EKvqrurak1VrQc2A9+rqo/0DdsB3NJd7XIFcKyqDvdvS5K0eE772xaT3AZQVduAncD1wCHgTeDWiaSTJI1srEKvqqeBp7vpbT3LC7hjksEkSePxTlFJaoSFLkmNsNAlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiNGeUj0eUm+n2RfkgNJPjNgzNVJjiXZ273uXZy4kqSFjPLEol8C11TV8STLgeeSPFFVu/rGPVtVN04+oiRpFEMLvXu83PFudnn3qsUMJUka30jn0JMsS7IXOAI8WVXPDxh2ZXda5okkF00ypCRpuJEKvap+XVWXAmuAjUku7huyB7iwqi4BHgAeHbSdJFuS7E6y++jRo6efWpL0NmNd5VJVPweeBjb1LX+jqo530zuB5UlWDnj/9qqararZmZmZ0w4tSXq7Ua5ymUmyopt+N3At8HLfmAuSpJve2G339YmnlSQtaJSrXFYBX0myjPmi/kZVPZbkNoCq2gZ8CLg9yQngLWBz92GqJOksGeUql/3AZQOWb+uZ3gpsnWw0SdI4vFNUkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGjHKM0XPS/L9JPuSHEjymQFjkuT+JIeS7E9y+eLElSQtZJRniv4SuKaqjidZDjyX5Imq2tUz5jpgQ/f6APBg909J0lky9Ai95h3vZpd3r/4HQN8MPNyN3QWsSLJqslElSacy0jn0JMuS7AWOAE9W1fN9Q1YDr/bMz3XL+rezJcnuJLuPHj16mpElSYOMVOhV9euquhRYA2xMcnHfkAx624DtbK+q2aqanZmZGTusJGlhY13lUlU/B54GNvWtmgPW9syvAV47k2CSpPGMcpXLTJIV3fS7gWuBl/uG7QBu6a52uQI4VlWHJx1WkrSwUa5yWQV8Jcky5v8H8I2qeizJbQBVtQ3YCVwPHALeBG5dpLySpAUMLfSq2g9cNmD5tp7pAu6YbDRJ0ji8U1SSGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaMcozRdcmeSrJwSQHktw5YMzVSY4l2du97l2cuJKkhYzyTNETwCeqak+S84EXkzxZVT/qG/dsVd04+YiSpFEMPUKvqsNVtaeb/gVwEFi92MEkSeMZ6xx6kvXMPzD6+QGrr0yyL8kTSS5a4P1bkuxOsvvo0aPjp5UkLWjkQk/yHuCbwMer6o2+1XuAC6vqEuAB4NFB26iq7VU1W1WzMzMzpxlZkjTISIWeZDnzZf5IVX2rf31VvVFVx7vpncDyJCsnmlSSdEqjXOUS4EvAwar63AJjLujGkWRjt93XJxlUknRqo1zlchXwUeCHSfZ2yz4FrAOoqm3Ah4Dbk5wA3gI2V1VNPq4kaSFDC72qngMyZMxWYOukQkmSxuedopLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktSIUZ4pujbJU0kOJjmQ5M4BY5Lk/iSHkuxPcvnixJUkLWSUZ4qeAD5RVXuSnA+8mOTJqvpRz5jrgA3d6wPAg90/JUlnydAj9Ko6XFV7uulfAAeB1X3DbgYernm7gBVJVk08rSRpQaMcof9WkvXAZcDzfatWA6/2zM91yw73vX8LsAVg3bp1Y0Z9Z1t/1+PTjvCOMq2f9yv33TCV/aoNI38omuQ9wDeBj1fVG/2rB7yl3ragantVzVbV7MzMzHhJJUmnNFKhJ1nOfJk/UlXfGjBkDljbM78GeO3M40mSRjXKVS4BvgQcrKrPLTBsB3BLd7XLFcCxqjq8wFhJ0iIY5Rz6VcBHgR8m2dst+xSwDqCqtgE7geuBQ8CbwK0TTypJOqWhhV5VzzH4HHnvmALumFQoSdL4vFNUkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGjHKM0UfSnIkyUsLrL86ybEke7vXvZOPKUkaZpRnin4Z2Ao8fIoxz1bVjRNJJEk6LUOP0KvqGeBnZyGLJOkMTOoc+pVJ9iV5IslFCw1KsiXJ7iS7jx49OqFdS5JgMoW+B7iwqi4BHgAeXWhgVW2vqtmqmp2ZmZnAriVJJ51xoVfVG1V1vJveCSxPsvKMk0mSxnLGhZ7kgiTppjd223z9TLcrSRrP0KtcknwNuBpYmWQO+DSwHKCqtgEfAm5PcgJ4C9hcVbVoiSVJAw0t9Kr68JD1W5m/rFGSNEXeKSpJjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNGFroSR5KciTJSwusT5L7kxxKsj/J5ZOPKUkaZpQj9C8Dm06x/jpgQ/faAjx45rEkSeMaWuhV9Qzws1MMuRl4uObtAlYkWTWpgJKk0Qx9SPQIVgOv9szPdcsO9w9MsoX5o3jWrVs3gV2ffevvenzaEaSJm+af61fuu2Fq+27NJD4UzYBlNWhgVW2vqtmqmp2ZmZnAriVJJ02i0OeAtT3za4DXJrBdSdIYJlHoO4BbuqtdrgCOVdXbTrdIkhbX0HPoSb4GXA2sTDIHfBpYDlBV24CdwPXAIeBN4NbFCitJWtjQQq+qDw9ZX8AdE0skSTot3ikqSY2w0CWpERa6JDXCQpekRljoktQIC12SGmGhS1IjLHRJaoSFLkmNsNAlqREWuiQ1wkKXpEZY6JLUCAtdkhphoUtSIyx0SWqEhS5JjRip0JNsSvLjJIeS3DVg/dVJjiXZ273unXxUSdKpjPJM0WXAF4APAnPAC0l2VNWP+oY+W1U3LkJGSdIIRjlC3wgcqqqfVNWvgK8DNy9uLEnSuEYp9NXAqz3zc92yflcm2ZfkiSQXDdpQki1JdifZffTo0dOIK0layCiFngHLqm9+D3BhVV0CPAA8OmhDVbW9qmaranZmZmasoJKkUxul0OeAtT3za4DXegdU1RtVdbyb3gksT7JyYiklSUONUugvABuSvDfJu4DNwI7eAUkuSJJuemO33dcnHVaStLChV7lU1YkkHwO+AywDHqqqA0lu69ZvAz4E3J7kBPAWsLmq+k/LSJIW0dBCh9+eRtnZt2xbz/RWYOtko0mSxuGdopLUCAtdkhphoUtSIyx0SWqEhS5JjbDQJakRFrokNcJCl6RGWOiS1AgLXZIaYaFLUiMsdElqhIUuSY2w0CWpERa6JDXCQpekRljoktSIkQo9yaYkP05yKMldA9Ynyf3d+v1JLp98VEnSqQwt9CTLgC8A1wHvAz6c5H19w64DNnSvLcCDE84pSRpilCP0jcChqvpJVf0K+Dpwc9+Ym4GHa94uYEWSVRPOKkk6hVEeEr0aeLVnfg74wAhjVgOHewcl2cL8ETzA8SQ/Hivt/1kJ/PQ037uYlmouWLrZzNUj/zR0SHM/rxH+nc9Ecz8v4MKFVoxS6BmwrE5jDFW1Hdg+wj5PHSjZXVWzZ7qdSVuquWDpZjPXeMw1nndarlFOucwBa3vm1wCvncYYSdIiGqXQXwA2JHlvkncBm4EdfWN2ALd0V7tcARyrqsP9G5IkLZ6hp1yq6kSSjwHfAZYBD1XVgSS3deu3ATuB64FDwJvArYsXGZjAaZtFslRzwdLNZq7xmGs876hcqXrbqW5J0jnIO0UlqREWuiQ14pwv9CR/l6SSrJx2FoAk/9B9/cHeJN9N8ofTzgSQ5LNJXu6yfTvJimlnAkjyl0kOJPlNkqlfXjbsay6mJclDSY4keWnaWU5KsjbJU0kOdv8N75x2JoAk5yX5fpJ9Xa7PTDtTryTLkvwgyWOT3vY5XehJ1gIfBP5r2ll6fLaq3l9VlwKPAfdOOc9JTwIXV9X7gf8A7p5ynpNeAv4CeGbaQUb8motp+TKwadoh+pwAPlFVfwxcAdyxRH5evwSuqapLgEuBTd3Vd0vFncDBxdjwOV3owD8Dn2TATUzTUlVv9Mz+HkskW1V9t6pOdLO7mL9XYOqq6mBVne4dw5M2ytdcTEVVPQP8bNo5elXV4ara003/gvmSWj3dVNB9BcnxbnZ591oSv4dJ1gA3AF9cjO2fs4We5Cbgv6tq37Sz9Evyj0leBf6KpXOE3utvgCemHWIJWugrLDREkvXAZcDzU44C/Pa0xl7gCPBkVS2JXMDnmT8I/c1ibHyUW/+nJsm/ARcMWHUP8Cngz85uonmnylVV/1pV9wD3JLkb+Bjw6aWQqxtzD/N/VX7kbGQaNdcSMdJXWOj/S/Ie4JvAx/v+hjo1VfVr4NLus6JvJ7m4qqb6+UOSG4EjVfVikqsXYx9LutCr6tpBy5P8CfBeYF8SmD99sCfJxqr6n2nlGuBfgMc5S4U+LFeSvwZuBP60zuINCGP8vKbNr7AYU5LlzJf5I1X1rWnn6VdVP0/yNPOfP0z7A+WrgJuSXA+cB/x+kq9W1UcmtYNz8pRLVf2wqv6gqtZX1XrmfxEvPxtlPkySDT2zNwEvTytLrySbgL8HbqqqN6edZ4ka5Wsu1Mn80dSXgINV9blp5zkpyczJq7iSvBu4liXwe1hVd1fVmq6zNgPfm2SZwzla6EvcfUleSrKf+VNCS+JSLmArcD7wZHdJ5bZpBwJI8udJ5oArgceTfGdaWboPjU9+zcVB4BtVdWBaeXol+Rrw78AfJZlL8rfTzsT8EedHgWu6P1N7u6PPaVsFPNX9Dr7A/Dn0iV8iuBR5678kNcIjdElqhIUuSY2w0CWpERa6JDXCQpekRljoktQIC12SGvG/bWiqs7IzT20AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "genes_with_peak = genes4[~pd.isna(genes4['fold_enrichment'])]\n",
    "plt.hist(genes_with_peak['log2FoldChange'], range=(-4,4))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f73cc6b7-3a35-454c-bf03-284374549a45",
   "metadata": {},
   "source": [
    "This looks differentbut is it significant?\n",
    "\n",
    "We'll use the `mannwhitneyu` function from `scipy.stats` to run a statistical test. To do this, we'll need to remove the `NaN` values using the `.dropna()` method."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "8ddf23bd-97f5-4162-bc0b-e8f160689259",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "MannwhitneyuResult(statistic=58374.0, pvalue=0.0014616513606629151)"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import scipy.stats as stats\n",
    "\n",
    "nochip = genes4[pd.isna(genes4['fold_enrichment'])]['log2FoldChange'].dropna()\n",
    "yeschip = genes4[~pd.isna(genes4['fold_enrichment'])]['log2FoldChange'].dropna()\n",
    "\n",
    "stats.mannwhitneyu(nochip, yeschip)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c761fe1b-ed73-4e19-a8a5-6a9d53f22d95",
   "metadata": {},
   "source": [
    "Finally, we'll make a table of the high-confidence targets that have a ChIP-Seq peak and an expression change."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "98d8ce3e-e07a-4108-85b9-e5833caff1bc",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqid</th>\n",
       "      <th>source</th>\n",
       "      <th>type</th>\n",
       "      <th>start_x</th>\n",
       "      <th>end_x</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>phase</th>\n",
       "      <th>attributes</th>\n",
       "      <th>name_x</th>\n",
       "      <th>...</th>\n",
       "      <th>-log10(pvalue)</th>\n",
       "      <th>fold_enrichment</th>\n",
       "      <th>-log10(qvalue)</th>\n",
       "      <th>name_y</th>\n",
       "      <th>baseMean</th>\n",
       "      <th>log2FoldChange</th>\n",
       "      <th>lfcSE</th>\n",
       "      <th>stat</th>\n",
       "      <th>pvalue</th>\n",
       "      <th>padj</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>349</th>\n",
       "      <td>chrII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>443821</td>\n",
       "      <td>444693</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YBR101C;Name=YBR101C;gene=FES1;Alias=FES1;O...</td>\n",
       "      <td>YBR101C</td>\n",
       "      <td>...</td>\n",
       "      <td>158.552</td>\n",
       "      <td>10.64490</td>\n",
       "      <td>154.402</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_69</td>\n",
       "      <td>3528.512278</td>\n",
       "      <td>-1.215401</td>\n",
       "      <td>0.289295</td>\n",
       "      <td>-4.201254</td>\n",
       "      <td>2.654410e-05</td>\n",
       "      <td>1.056289e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1243</th>\n",
       "      <td>chrIV</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>892875</td>\n",
       "      <td>893927</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YDR214W;Name=YDR214W;gene=AHA1;Alias=AHA1;O...</td>\n",
       "      <td>YDR214W</td>\n",
       "      <td>...</td>\n",
       "      <td>117.605</td>\n",
       "      <td>9.47421</td>\n",
       "      <td>113.817</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_162</td>\n",
       "      <td>3436.341557</td>\n",
       "      <td>-1.875259</td>\n",
       "      <td>0.301106</td>\n",
       "      <td>-6.227899</td>\n",
       "      <td>4.727310e-10</td>\n",
       "      <td>3.762348e-07</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1288</th>\n",
       "      <td>chrIV</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>971808</td>\n",
       "      <td>974243</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YDR258C;Name=YDR258C;gene=HSP78;Alias=HSP78...</td>\n",
       "      <td>YDR258C</td>\n",
       "      <td>...</td>\n",
       "      <td>130.964</td>\n",
       "      <td>9.92330</td>\n",
       "      <td>127.060</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_167</td>\n",
       "      <td>1290.992019</td>\n",
       "      <td>-1.464184</td>\n",
       "      <td>0.325875</td>\n",
       "      <td>-4.493090</td>\n",
       "      <td>7.019698e-06</td>\n",
       "      <td>3.438032e-03</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2474</th>\n",
       "      <td>chrVII</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>772454</td>\n",
       "      <td>773686</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YGR142W;Name=YGR142W;gene=BTN2;Alias=BTN2;O...</td>\n",
       "      <td>YGR142W</td>\n",
       "      <td>...</td>\n",
       "      <td>186.912</td>\n",
       "      <td>10.29620</td>\n",
       "      <td>182.373</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_372</td>\n",
       "      <td>420.520506</td>\n",
       "      <td>-1.922651</td>\n",
       "      <td>0.368085</td>\n",
       "      <td>-5.223394</td>\n",
       "      <td>1.756733e-07</td>\n",
       "      <td>1.016829e-04</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5382</th>\n",
       "      <td>chrXIV</td>\n",
       "      <td>SGD</td>\n",
       "      <td>gene</td>\n",
       "      <td>618507</td>\n",
       "      <td>619565</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>ID=YNL007C;Name=YNL007C;gene=SIS1;Alias=SIS1,t...</td>\n",
       "      <td>YNL007C</td>\n",
       "      <td>...</td>\n",
       "      <td>186.175</td>\n",
       "      <td>9.87451</td>\n",
       "      <td>181.642</td>\n",
       "      <td>Hsf1_ChIP_heatshk_peak_708</td>\n",
       "      <td>1067.866637</td>\n",
       "      <td>-2.608539</td>\n",
       "      <td>0.292907</td>\n",
       "      <td>-8.905696</td>\n",
       "      <td>5.305227e-19</td>\n",
       "      <td>6.755676e-16</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  29 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "       seqid source  type  start_x   end_x score strand phase  \\\n",
       "349    chrII    SGD  gene   443821  444693     .      -     .   \n",
       "1243   chrIV    SGD  gene   892875  893927     .      +     .   \n",
       "1288   chrIV    SGD  gene   971808  974243     .      -     .   \n",
       "2474  chrVII    SGD  gene   772454  773686     .      +     .   \n",
       "5382  chrXIV    SGD  gene   618507  619565     .      -     .   \n",
       "\n",
       "                                             attributes   name_x  ...  \\\n",
       "349   ID=YBR101C;Name=YBR101C;gene=FES1;Alias=FES1;O...  YBR101C  ...   \n",
       "1243  ID=YDR214W;Name=YDR214W;gene=AHA1;Alias=AHA1;O...  YDR214W  ...   \n",
       "1288  ID=YDR258C;Name=YDR258C;gene=HSP78;Alias=HSP78...  YDR258C  ...   \n",
       "2474  ID=YGR142W;Name=YGR142W;gene=BTN2;Alias=BTN2;O...  YGR142W  ...   \n",
       "5382  ID=YNL007C;Name=YNL007C;gene=SIS1;Alias=SIS1,t...  YNL007C  ...   \n",
       "\n",
       "      -log10(pvalue)  fold_enrichment -log10(qvalue)  \\\n",
       "349          158.552         10.64490        154.402   \n",
       "1243         117.605          9.47421        113.817   \n",
       "1288         130.964          9.92330        127.060   \n",
       "2474         186.912         10.29620        182.373   \n",
       "5382         186.175          9.87451        181.642   \n",
       "\n",
       "                          name_y     baseMean  log2FoldChange     lfcSE  \\\n",
       "349    Hsf1_ChIP_heatshk_peak_69  3528.512278       -1.215401  0.289295   \n",
       "1243  Hsf1_ChIP_heatshk_peak_162  3436.341557       -1.875259  0.301106   \n",
       "1288  Hsf1_ChIP_heatshk_peak_167  1290.992019       -1.464184  0.325875   \n",
       "2474  Hsf1_ChIP_heatshk_peak_372   420.520506       -1.922651  0.368085   \n",
       "5382  Hsf1_ChIP_heatshk_peak_708  1067.866637       -2.608539  0.292907   \n",
       "\n",
       "          stat        pvalue          padj  \n",
       "349  -4.201254  2.654410e-05  1.056289e-02  \n",
       "1243 -6.227899  4.727310e-10  3.762348e-07  \n",
       "1288 -4.493090  7.019698e-06  3.438032e-03  \n",
       "2474 -5.223394  1.756733e-07  1.016829e-04  \n",
       "5382 -8.905696  5.305227e-19  6.755676e-16  \n",
       "\n",
       "[5 rows x 29 columns]"
      ]
     },
     "execution_count": 42,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "targets = genes4[(~pd.isna(genes4['fold_enrichment']))\n",
    "                 & (genes4['padj'] < 0.1)]\n",
    "targets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0b8a667-b286-456f-b53f-a4ae9f8427f8",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
